File: 101/linsolve.c
#include <stdlib.h>
#include <gmp.h>
#include "linsolve.h"

int linsolve(size_t nxs, mpq_t **eqns, mpq_t *xs){
	mpq_t prod;
	mpq_init(prod);
	for(size_t c = 0; c < nxs; ++c){
		if(!mpz_cmp_ui(mpq_numref(eqns[c][c]), 0)){//this row starts with zero; try switching it with one that does not
			mpq_t *t;
			for(size_t i = c + 1; i < nxs; ++i){
				if(mpz_cmp_ui(mpq_numref(eqns[i][c]), 0)){
					t = eqns[i];
					eqns[i] = eqns[c];
					eqns[c] = t;
					goto SWITCHED_ROWS;
				}
			}
			mpq_clear(prod);
			return 0;
			SWITCHED_ROWS:;
		}
		for(size_t j = c; j <= nxs; ++j){//for each element
			//divide eqns[c][j] by eqns[c][c]
			mpq_div(eqns[c][j], eqns[c][j], eqns[c][c]);
		}
		for(size_t i = 0; i < nxs; ++i){//for each row
			if(i == c){
				continue;
			}
			for(size_t j = c; j <= nxs; ++j){//for each element
				//subtract eqns[c][j]*eqns[i][c] from eqns[i][j]
				mpq_mul(prod, eqns[c][j], eqns[i][c]);
				mpq_sub(eqns[i][j], eqns[i][j], prod);
			}
		}
	}
	mpq_clear(prod);
	for(size_t i = 0; i < nxs; ++i){
		mpq_set(xs[i], eqns[i][nxs]);
	}
	return 1;
}

File: 101/linsolve.py
def linsolve(eqns):
	n, m = len(eqns), len(eqns[0])
	for c in range(n):
		if eqns[c][c] == 0:
			for s in range(c + 1, n):
				if eqns[s][c]:
					eqns[c], eqns[s] = eqns[s], eqns[c]
					break
			else:
				return None
		for j in range(c, m)[::-1]:
			eqns[c][j] /= eqns[c][c]
		for i in range(n):
			if i == c:
				continue
			for j in range(c, m)[::-1]:
				eqns[i][j] -= eqns[c][j]*eqns[i][c]
	return [row[m - 1] for row in eqns]

def pval(ns, x):
	return sum([n*x**i for (i, n) in enumerate(ns)])

